lib/puppet/provider/package_targetable.rb

```diff
@@ -0,0 +1,55 @@
+# Targetable package providers implement a 'command' attribute.
+#
+# The packages hash passed to prefetch() is deduplicated, as it is keyed only by name (in transaction.rb).
+# The packages hash should be keyed by all namevars.
+# Instead ...
+#
+# Inspect any package to access the catalog (every package includes a reference to the catalog).
+# Inspect the catalog to find all of the commands for all of the packages of this class.
+# Find all of the package instances using each command including the default of the provider command.
+# Assign each instance's provider by selecting it from the packages hash, based upon name and command,
+
+class Puppet::Provider::Package::Targetable < Puppet::Provider::Package
+  # Prefetch our package list, yo.
+  def self.prefetch(packages)
+    catalog_packages = packages.first[1]::catalog::resources.select{ |p| p.provider.class == self }
```

Comment:
It would be nice to have a chat with coremunity about whether we want to expose a prefetchv2 or something that can take a better data structure here.

Reply:
Assuming this is where prefetch (v1) calls `provider_class.prefetch(resources)`

https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/transaction.rb#L360

I would not know who to ask, what to propose :)

Reply:
@branan Met with Coremunity. If we cannot move forward with a prefetchV2 mechanism sooner than later, I've updated the above code comments to better document the need and to document the interest for implementation later.

@jhelwig If we are concerned about implementing these changes to gem and pip in Puppet Core, implementing `targetable` in the package type and provider by merging only the `3131777` commit would allow users (me) to implement this functionality via a Forge module(s) that inherits from a base package provider(s)?